src/main/java/rocks/cleanstone/game/world/chunk/ProximityNearbyChunkRetriever.java

```diff
@@ -0,0 +1,57 @@
+package rocks.cleanstone.game.world.chunk;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.scheduling.annotation.AsyncResult;
+import org.springframework.stereotype.Service;
+import org.springframework.util.concurrent.ListenableFuture;
+import rocks.cleanstone.game.world.World;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+@Service
+public class ProximityNearbyChunkRetriever implements NearbyChunkRetriever {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Override
+    public Collection<ChunkCoords> getChunkCoordsAround(int chunkX, int chunkZ, int radius) {
+        Collection<ChunkCoords> coords = new ArrayList<>();
+        coords.add(ChunkCoords.of(chunkX, chunkZ));
+
+        // generate positions around chunk in order of proximity
+        for (int distance = 1; distance <= radius * 1.5; distance++) {
+            for (int relZ = Math.max(0, distance - radius); relZ < Math.min(distance, radius + 1); relZ++) {
+                int relX = distance - relZ;
+                coords.add(ChunkCoords.of(chunkX + relX, chunkZ + relZ));
+                coords.add(ChunkCoords.of(chunkX + relZ, chunkZ - relX));
+                coords.add(ChunkCoords.of(chunkX - relX, chunkZ - relZ));
+                coords.add(ChunkCoords.of(chunkX - relZ, chunkZ + relX));
+            }
+        }
+        return coords;
+    }
+
+    @Async
+    @Override
+    public ListenableFuture<Collection<Chunk>> getChunksAround(int chunkX, int chunkZ, int radius, World world) {
+        try {
+            return new AsyncResult<>(getChunkCoordsAround(chunkX, chunkZ, radius).stream()
+                    .map(coord -> world.getChunk(coord.getX(), coord.getZ()))
+                    .map(chunkFuture -> {
+                        try {
+                            return chunkFuture.get();
+                        } catch (InterruptedException | ExecutionException e) {
+                            logger.error("Failed to get chunk " + chunkX + ":" + chunkZ + " in world "
+                                    + world.getWorldConfig().getName(), e);
+                            throw new RuntimeException(e);
+                        }
+                    }).collect(Collectors.toSet()));
```

Comment:
Warum ein Set? Das führt beim collecten zu vielen contains checks.

Reply:
Was würdest du sonst nehmen? HashSet ist allgemein eigentlich schneller als ArrayList mit einem O(1) beim contains().

Reply:
Beim collecten in eine List laufen gar keine contains checks, beim bauen eines sets schon weil es ja Duplikate vermeiden muss. Und das keine chunks doppelt drin sind ist ja durch `getChunkCoordsAround` gegeben.

Reply:
Ja aber toList returnt eine ArrayList und kein HashSet obwohl ein HashSet für die Daten besser geeignet wäre da Chunks in der Collection nur einmal vorkommen werden. Beim collecten wäre ein toList zwar schneller aber für jedes weitere contains() des Methodenaufrufers wäre ein HashSet deutlich schneller. Das contains() beim collecten ist durch sein O(1) doch sowieso kaum ein Overhead im Vergleich zum contains() bei einer ArrayList das durch den Methodenaufrufer durchaus direkt oder indirekt verwendet werden könnte.

Reply:
Ja, stimmt. Ich habe bis jetzt nur an das collect selbst gedacht, und nicht daran was danach noch mit den Daten passieren könnte. Mit `toSet` ist es dann besser!